pkg/releaselock/controller.go

```diff
@@ -94,114 +117,232 @@ func NewLockReleaseController(client kubernetes.Interface, config *LockReleaseCo
-			if err != nil {
-				klog.Errorf("Failed to list nodes: %v", err)
-				return
-			}
-			klog.Infof("Listed %d nodes", len(nodes))
-
-			for _, cm := range cmList.Items {
-				// Filter out root ca.
-				if cm.Name == rootCA {
-					continue
-				}
-				if err := c.syncLockInfo(ctx, &cm, nodes); err != nil {
-					klog.Errorf("Failed to sync lock info for configmap %s/%s: %v", cm.Namespace, cm.Name, err)
-				}
-			}
-		}, c.config.SyncPeriod)
-	}
-
-	rl, err := resourcelock.New(
-		resourcelock.LeasesResourceLock,
-		util.ManagedFilestoreCSINamespace,
-		leaseName,
-		nil,
-		c.client.CoordinationV1(),
-		resourcelock.ResourceLockConfig{
-			Identity: c.id,
-		})
-	if err != nil {
-		klog.Fatalf("Error creating resourcelock: %v", err)
+func (c *LockReleaseController) Run(ctx context.Context) error {
+	defer utilruntime.HandleCrash()
+	defer c.updateEventQueue.ShutDown()
+	defer c.createEventQueue.ShutDown()
+	klog.Infof("Lock release controller %s started leading on node %s", c.GetId(), c.GetHost())
+	if !cache.WaitForCacheSync(ctx.Done(), (*c.nodeInformer).HasSynced) {
+		klog.Fatal("Timed out waiting for caches to sync")
+		return fmt.Errorf("failed to wait for caches to sync")
 	}
+	klog.Info("Cache sync completed successfully.")
+	go wait.UntilWithContext(ctx, c.runCreateEventWorker, time.Second)
+	go wait.UntilWithContext(ctx, c.runUpdateEventWorker, time.Second)
+	klog.Info("Started workers")
+	<-ctx.Done()
+	klog.Info("Shutting down workers")
+	return nil
+}
 
-	leaderelection.RunOrDie(ctx, leaderelection.LeaderElectionConfig{
-		Lock:          rl,
-		LeaseDuration: c.config.LeaseDuration,
-		RenewDeadline: c.config.RenewDeadline,
-		RetryPeriod:   c.config.RetryPeriod,
-		Callbacks: leaderelection.LeaderCallbacks{
-			OnStartedLeading: run,
-			OnStoppedLeading: func() {
-				klog.Fatalf("%s no longer the leader", c.id)
-			},
-		},
-	})
+func (c *LockReleaseController) runCreateEventWorker(ctx context.Context) {
+	for c.processNextCreateEvent(ctx) {
+	}
 }
 
-func (c *LockReleaseController) syncLockInfo(ctx context.Context, cm *corev1.ConfigMap, nodes map[string]*corev1.Node) error {
-	nodeName, err := GKENodeNameFromConfigMap(cm)
+// TODO(b/374327452): interface rpc calls for mocking and create unit tests for handleCreateEvent and handleUpdateEvent
+func (c *LockReleaseController) handleCreateEvent(ctx context.Context, obj interface{}) error {
+	node := obj.(*corev1.Node)
+	start := time.Now()
+	cmName := ConfigMapNamePrefix + node.Name
+	cm, err := c.client.CoreV1().ConfigMaps(util.ManagedFilestoreCSINamespace).Get(ctx, cmName, metav1.GetOptions{})
+	duration := time.Since(start)
+	c.RecordKubeAPIMetrics(err, metrics.ConfigMapResourceType, metrics.GetOpType, metrics.ReconcilerOpSource, duration)
+
 	if err != nil {
-		klog.Errorf("Failed to get GKE node name from configmap %s/%s: %v", cm.Namespace, cm.Name, err)
+		klog.Errorf("Failed to get configmap in namespace %s: %v", util.ManagedFilestoreCSINamespace, err)
```

Comment:
I believe we can categorize the errors as follows:
1. If it's an API server error, we should requeue the request for a retry.
2. If the ConfigMap is missing, it's a no-op, and we can safely skip it.
For reference, https://github.com/kubernetes-sigs/gcp-filestore-csi-driver/blob/master/pkg/releaselock/configmap_util.go#L104-L113

Reply:
done in next revision